README.md
# Muta benchmark

## DEPRECATED

move to https://github.com/HuobiGroup/huobi-chain-benchmark

## Requirement

- NodeJS >= 10

## Quick start

```shell
git clone https://github.com/nervosnetwork/muta-benchmark.git
cd muta-benchmark
npm install
node index -d 5 http://127.0.0.1:8000/graphql
```

or install globally

```
git clone https://github.com/nervosnetwork/muta-benchmark.git
cd muta-benchmark
npm install
[sudo] npm link
muta-bench -d 5 http://127.0.0.1:8000/graphql
```

## Options

```
Options:
  -m --method [method]            HTTP method (default: "POST")
  -d --duration [duration]        number of second (default: 5)
  -g --gap [gap]                  muta's timeout_gap (default: 20)
  -c --connections [connections]  number of connection (default: 20)
  --pk [privateKey]               sender of the transfer
  --receiver [receiver]           receiver of the transfer
  --verbose                       show verbose info, use it for debug (default: false)
  -h, --help                      output usage information
```

## Q&A

### What can I do when unknown error appeared?

Try to use `--verbose` for more bench information

### How to customize private key of the sender?

- use environment variable `PRIVATE_KEY=0x...`
- or use `--pk 0x...` option

### How to run with multiple nodes

```
# each node will be allocated 15 connections
muta-bench -c 30 http://127.0.0.1:8000/graphql http://127.0.0.1:8001/graphql
```

It is recommended to adjust the number of connections to n times the number of nodes


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.3](https://github.com/homura/huobi-chain-benchmark/compare/v0.2.1...v0.2.3) (2020-08-03)


### Features

* adapting huobi-chain@0.4.0-alpha ([361683d](https://github.com/homura/huobi-chain-benchmark/commit/361683d92de359c54cb573d3dedea0ea6e649981))
* supported miner in asset service ([eb9d60a](https://github.com/homura/huobi-chain-benchmark/commit/eb9d60ab9b447a2d02181fa5be523cf81991dc0c))
* update muta-sdk ([1d443ca](https://github.com/homura/huobi-chain-benchmark/commit/1d443caffc9cad61edbade684d9d06b4e9c94afc))
* update sdk ([7a2acb9](https://github.com/homura/huobi-chain-benchmark/commit/7a2acb9157c674db7ab7ff522a64bf4e15171744))
* update sdk version to 0.10 ([1815245](https://github.com/homura/huobi-chain-benchmark/commit/1815245087ab981bc6c90cf019f1235b9a7ddd34))
* update to huobi-chain@0.5.0-beta.1 ([7bd570a](https://github.com/homura/huobi-chain-benchmark/commit/7bd570ac1f5322d6d10d67a2607e5c26ca10389d))


### Bug Fixes

* fix get balance error ([a2c4db3](https://github.com/homura/huobi-chain-benchmark/commit/a2c4db3e60ca115ac857d9c6d31b6b0b6509a2b5))

### [0.2.2](https://github.com/nervosnetwork/muta-benchmark/compare/v0.2.1...v0.2.2) (2020-07-14)


### Features

* update muta-sdk ([1d443ca](https://github.com/nervosnetwork/muta-benchmark/commit/1d443caffc9cad61edbade684d9d06b4e9c94afc))
* update sdk ([7a2acb9](https://github.com/nervosnetwork/muta-benchmark/commit/7a2acb9157c674db7ab7ff522a64bf4e15171744))
* update sdk version to 0.10 ([1815245](https://github.com/nervosnetwork/muta-benchmark/commit/1815245087ab981bc6c90cf019f1235b9a7ddd34))


### Bug Fixes

* fix get balance error ([a2c4db3](https://github.com/nervosnetwork/muta-benchmark/commit/a2c4db3e60ca115ac857d9c6d31b6b0b6509a2b5))

### [0.2.1](https://github.com/homura/muta-benchmark/compare/v0.1.7...v0.2.1) (2020-06-02)

## [0.2.0](https://github.com/homura/muta-benchmark/compare/v0.1.4...v0.2.0) (2020-03-13)


### Features

* supported precision for AssetService ([a0bf45c](https://github.com/homura/muta-benchmark/commit/a0bf45c6a4d6bbe78b7ee7a21cc75880d4749c30))

### [0.1.4](https://github.com/homura/muta-benchmark/compare/v0.1.3...v0.1.4) (2020-03-10)


### Bug Fixes

* compatible with millisecond and second timestamp ([71bdabd](https://github.com/homura/muta-benchmark/commit/71bdabd33fdd9ad653a1b1c2132b3c1e0ad6b1d0))

### [0.1.3](https://github.com/homura/muta-benchmark/compare/v0.1.2...v0.1.3) (2020-03-10)


### Features

* force hex string to start with 0x ([c915ab6](https://github.com/homura/muta-benchmark/commit/c915ab6e5aadee9653eea81e05c97e92c461d4fa))
* supported [millis timestamp](https://github.com/nervosnetwork/muta/pull/216) ([ea3b25f](https://github.com/homura/muta-benchmark/commit/ea3b25f655cc00779d097a7e53bacc52b4e52724))

### [0.1.2](https://github.com/homura/muta-benchmark/compare/v0.1.1...v0.1.2) (2020-02-26)


### Bug Fixes

* fix failure caused by custom `gap` or `chain-id` ([c959b25](https://github.com/homura/muta-benchmark/commit/c959b254ed58555daf2e1d5ddfbc9b7fd9bccafe))

### [0.1.1](https://github.com/homura/muta-benchmark/compare/v0.1.0...v0.1.1) (2020-02-23)


### Features

* add json output ([#5](https://github.com/homura/muta-benchmark/issues/5)) ([a2ac419](https://github.com/homura/muta-benchmark/commit/a2ac4190a55d703af8f4535f6e3c6bc4ede4958c))
* pretty json output ([e3a71ce](https://github.com/homura/muta-benchmark/commit/e3a71ceaf3cf03babe42e7983a61deed23c776c8))

## [0.1.0](https://github.com/homura/muta-benchmark/compare/v0.0.2...v0.1.0) (2020-02-22)


### Bug Fixes

* Add Content-type to http.headers ([#4](https://github.com/homura/muta-benchmark/issues/4)) ([5a8a911](https://github.com/homura/muta-benchmark/commit/5a8a9118bc45554464a21cfc651789150faf63bc))

### 0.0.2 (2020-02-14)


### Features

* output round usage when TPS ([aea7c53](https://github.com/homura/muta-benchmark/commit/aea7c53f677ddc4ed2356e8edda56d614d13e0e1))
* Support cluster ([b50a031](https://github.com/homura/muta-benchmark/commit/b50a03109fd1f3d1e7a71ddf454c1752f3f483cd))
* supported max tx count per second ([5ed35ee](https://github.com/homura/muta-benchmark/commit/5ed35eed64308311a171540194504f7cf702f6df))
* supported prepare signature ([77618b0](https://github.com/homura/muta-benchmark/commit/77618b005d4c34ee5311c37d28648f856ebf3f04))
* tip error when create asset failed ([ecf403f](https://github.com/homura/muta-benchmark/commit/ecf403f97ea85a9b57eeb69d33441d26be4c3141))
* update muta-sdk ([6e0d0f6](https://github.com/homura/muta-benchmark/commit/6e0d0f6d06844879a82d2255e4b972193ce3029d))
* update sdk ([d6be868](https://github.com/homura/muta-benchmark/commit/d6be868549072508c5999f1c958ae4c220a7c7e0))


### Bug Fixes

* fix the calculation results error of TPS ([abee51e](https://github.com/homura/muta-benchmark/commit/abee51e6b939aebaa015def95ce40e909d768250))


bench.js
const autocannon = require('@homura/autocannon');
const ora = require('ora');
const Table = require('cli-table3');
const logger = require('./logger');
const { utils } = require('@mutadev/muta-sdk');
const randomBytes = require('randombytes');

function round(x) {
  return parseFloat(Math.round(x * 100) / 100).toFixed(2);
}

const query = `mutation ( $inputRaw: InputRawTransaction! $inputEncryption: InputTransactionEncryption! ) { sendTransaction(inputRaw: $inputRaw, inputEncryption: $inputEncryption) }`;

async function runMain(assetBenchProducer, workers, options) {
  let errorCount = 0;

  function getBody() {
    return assetBenchProducer.produceRequestBody();
  }

  await assetBenchProducer.start();

  return new Promise((resolve, reject) => {
    const instance = autocannon(
      {
        ...options,
        setupClient(client) {
          client.setBody(getBody());
        },
      },
      finishedBench
    );

    autocannon.track(instance);

    instance.on('response', function (
      client,
      statusCode,
      returnBytes,
      responseTime
    ) {
      const res = client.parser.chunk.toString();
      const isError = res.includes('error');
      if (isError) {
        logger.error(res);
        errorCount++;
      }
      client.setBody(getBody());
    });

    instance.on('done', async function ({ start, duration }) {
      for (const worker of workers) {
        worker.process.kill();
      }
      const spin = ora('TPS is calculating ').start();
      const {
        blockUsage,
        transferProcessed,
        blocks,
      } = await assetBenchProducer.end();
      spin.stop();

      const txCount = transferProcessed;
      const blockCount = blockUsage;

      const balanceTable = new Table({ head: ['', 'balance', 'block height'] });
      balanceTable.push(
        {
          init: [
            assetBenchProducer.startBalance,
            assetBenchProducer.startBlock,
          ],
        },
        { done: [assetBenchProducer.endBalance, assetBenchProducer.endBlock] }
      );

      console.log('block_id \t\t count \t\t\t round');
      Object.entries(blocks)
        .sort((l, r) => Number(l[0]) - Number(r[0]))
        .forEach(([id, info]) => {
          console.log(
            `${id} \t\t\t ${info.transactionsCount} \t\t\t ${info.round}`
          );
        });

      console.log('TPS:');
      console.log(balanceTable.toString());

      console.log(`${round(txCount / blockCount)} tx/block`);
      console.log(`${round(duration / blockCount)} sec/block`);
      console.log(`${round(txCount / duration)} tx/sec`);

      if (options.json) {
        const benchDetail = {
          blocks: [],
          tx_block: round(txCount / blockCount),
          sec_block: round(duration / blockCount),
          tx_sec: round(txCount / duration),
        };
        Object.entries(blocks)
          .sort((l, r) => Number(l[0]) - Number(r[0]))
          .forEach(([id, info]) => {
            benchDetail.blocks.push([
              parseInt(id),
              info.transactionsCount,
              info.round,
            ]);
          });
        console.log(JSON.stringify(benchDetail));
      }
    });

    function finishedBench(err) {
      if (err) reject(err);
      else resolve({ errorCount });
    }
  });
}

function runWorker() {
  let errorCount = 0;

  const workerData = JSON.parse(process.env.WORKER_DATA);
  const options = JSON.parse(process.env.OPTIONS);

  const payload = JSON.stringify({
    asset_id: workerData.assetId,
    to: workerData.to,
    value: 1,
    memo: 'Hello',
  });

  function getBody() {
    const signed = utils.signTransaction(
      {
        serviceName: 'asset',
        method: 'transfer',
        payload,
        timeout: workerData.timeout,
        nonce: `0x${randomBytes(32).toString('hex')}`,
        chainId: `${workerData.chainId}`,
        cyclesPrice: '0x01',
        cyclesLimit: '0xffffffff',
        sender: workerData.sender,
      },
      Buffer.from(workerData.privateKey, 'hex')
    );

    const variables = {
      inputRaw: utils.separateOutRawTransaction(signed),
      inputEncryption: utils.separateOutEncryption(signed),
    };

    const tx = JSON.stringify({
      query,
      variables,
    });

    return tx;
  }

  const instance = autocannon(
    {
      ...options,
      setupClient(client) {
        client.setBody(getBody());
      },
    },
    finishedBench
  );

  autocannon.track(instance);

  instance.on('response', function (
    client,
    statusCode,
    returnBytes,
    responseTime
  ) {
    const res = client.parser.chunk.toString();
    const isError = res.includes('error');
    if (isError) {
      logger.error(res);
      errorCount++;
    }
    client.setBody(getBody());
  });

  function finishedBench(err) {
    if (err) {
      process.exit(1);
    }
    process.exit(0);
  }
}

exports.runWorker = runWorker;
exports.runMain = runMain;


BenchProducer.js
const { Muta, utils } = require('@mutadev/muta-sdk');
const { AssetService } = require('@mutadev/service');
const randomBytes = require('randombytes');

const query = `mutation ( $inputRaw: InputRawTransaction! $inputEncryption: InputTransactionEncryption! ) { sendTransaction(inputRaw: $inputRaw, inputEncryption: $inputEncryption) }`;

function hexToTimestamp(hex) {
  const timestamp = Number(utils.toHex(hex));
  if (timestamp <= 9999999999) {
    return timestamp * 1000;
  }
  return timestamp;
}

class AssetBench {
  constructor(options) {
    const { pk, chainId, gap, url, receiver } = options;
    const muta = new Muta({
      chainId,
      timeoutGap: gap,
      endpoint: Array.isArray(url) ? url[0] : url,
    });

    this.client = muta.client();
    this.account = Muta.accountFromPrivateKey(pk);
    this.privateKey = pk;
    this.rawClient = this.client.getRawClient();
    this.service = new AssetService(this.client, this.account);

    this.chainId = chainId;

    this.to = Muta.accountFromPrivateKey(randomBytes(32)).address;
    this.receiver = receiver;
    this.gap = gap;
  }

  async createAsset() {
    const asset = await this.service.write.create_asset({
      supply: 99999999,
      symbol: 'M' + (Math.random() * 1e8).toString().substring(0, 4),
      name: 'M' + (Math.random() * 1e8).toString().substring(0, 4),
      precision: 0,
      relayable: false,
      admin: this.account.address,
      init_mints: [{ addr: this.account.address, balance: 99999999 }],
    });
    this.assetId = asset.response.response.succeedData.id;
    await this.client.waitForNextNBlock(1);
    return asset;
  }

  async prepare() {
    const startBlock = await this.client.getLatestBlockHeight();
    this.timeout = utils.toHex(Number(startBlock + Number(this.gap) - 1));
  }

  async start() {
    this.startTime = Date.now();
    this.startBlock = await this.client.getLatestBlockHeight();
    this.startBalance = await this.service.read
      .get_balance({
        asset_id: this.assetId,
        user: this.account.address,
      })
      .then((res) => res.succeedData.balance);
  }

  async end() {
    this.endTime = Date.now();

    await this.client.waitForNextNBlock(1);

    this.endBlock = await this.client.getLatestBlockHeight();
    this.endBalance = await this.service.read
      .get_balance({
        asset_id: this.assetId,
        user: this.account.address,
      })
      .then((res) => res.succeedData.balance);

    const blocks = {};
    for (let height = this.startBlock; height <= this.endBlock; height++) {
      const res = await this.rawClient.getBlock({
        height: utils.toHex(height),
      });

      blocks[height] = {
        round: Number(res.getBlock.header.proof.round),
        timeStamp: hexToTimestamp(res.getBlock.header.timestamp),
        transactionsCount: res.getBlock.orderedTxHashes.length,
      };
    }

    return {
      timeUsage: this.endTime - this.startTime,
      // (- 1) is for wait 1 block to ensure all tx are finished
      blockUsage: this.endBlock - this.startBlock - 1,
      transferProcessed: this.startBalance - this.endBalance,

      blocks,
    };
  }

  produceRequestBody() {
    const timeout = this.timeout;
    const chainId = this.chainId;

    const assetId = this.assetId;
    const to = this.to;

    const signed = utils.signTransaction(
      {
        serviceName: 'asset',
        method: 'transfer',
        payload: JSON.stringify({
          asset_id: assetId,
          to: to,
          value: 1,
          memo: 'hello',
        }),
        timeout: timeout,
        nonce: `0x${randomBytes(32).toString('hex')}`,
        chainId: `${chainId}`,
        cyclesPrice: '0x01',
        cyclesLimit: '0xfffffff',
        sender: this.account.address,
      },
      utils.toBuffer(this.privateKey)
    );

    const variables = {
      inputRaw: utils.separateOutRawTransaction(signed),
      inputEncryption: utils.separateOutEncryption(signed),
    };

    return JSON.stringify({
      query,
      variables,
    });
  }
}

exports.AssetBenchProducer = AssetBench;


index.js
#!/usr/bin/env node

const program = require('commander');
const ora = require('ora');
const cluster = require('cluster');
const { utils, Account } = require('@mutadev/muta-sdk');
const { runMain, runWorker } = require('./bench');
const logger = require('./logger');
const { AssetBenchProducer } = require('./BenchProducer');

const toNumber = (v) => parseInt(v);

const receiverAccount = new Account(
  '0x0000000000000000000000000000000000000000000000000000000000000001'
);

const { args } = program
  .option('-m --method [method]', 'HTTP method', 'POST')
  .option('-d --duration [duration]', 'number of second', toNumber, 60)
  .option('-g --gap [gap]', "muta's timeout_gap", toNumber, 20)
  .option(
    '-c --connections [connections]',
    'number of connection',
    toNumber,
    20
  )
  .option(
    '--pk [privateKey]',
    'sender of the transfer',
    utils.toHex,
    process.env.PRIVATE_KEY ||
      '0x45c56be699dca666191ad3446897e0f480da234da896270202514a0e1a587c3f'
  )
  .option(
    '--chain-id [chainId]',
    'chain id',
    utils.toHex,
    '0xb6a4d7da21443f5e816e8700eea87610e6d769657d6b8ec73028457bf2ca4036'
  )
  .option(
    '--receiver [receiver]',
    'receiver of the transfer',
    utils.toHex,
    receiverAccount.address
  )
  .option('--verbose', 'show verbose info, use it for debug', false)
  .option('--cpu [cpu]', 'cpu nums', toNumber, 3)
  .option('--json', 'print output as json', false)
  .name('muta-bench')
  .usage(
    '-m POST -d 60 -c 20 --gap 30 --pk 0x45c56be699dca666191ad3446897e0f480da234da896270202514a0e1a587c3f http://127.0.0.1:8000/graphql'
  )
  .parse(process.argv);

const opts = program.opts();
const url = args.length ? args : 'http://127.0.0.1:8000/graphql';

opts.url = url;

if (opts.verbose) {
  logger.enable('*');
}

if (cluster.isMaster) {
  run().catch(({ errorCount }) => {
    if (errorCount > 0) {
      console.error(`HTTP success but GraphQL error count: ${errorCount}`);
    }
  });
} else {
  runWorker();
}

async function run() {
  console.log(opts);
  const { gap, pk, url, receiver, chainId } = opts;
  const assetBenchProducer = new AssetBenchProducer({
    pk,
    chainId,
    gap,
    url,
    receiver,
  });


  const createAssetSpin = ora('Creating asset').start();
  try {
    const asset = await assetBenchProducer.createAsset();
    createAssetSpin.succeed(`Created asset ${JSON.stringify(asset)}`);
  } catch (e) {
    createAssetSpin.fail(`Asset create failed, ${e.message}`);
  }

  await assetBenchProducer.prepare();

  opts.headers = {};
  opts.headers['Content-Type'] = 'application/json';

  const workers = Array.from({ length: opts.cpu }).map(() => {
    const worker = cluster.fork({
      WORKER_DATA: JSON.stringify({
        assetId: assetBenchProducer.assetId,
        to: assetBenchProducer.to,
        timeout: assetBenchProducer.timeout,
        chainId: assetBenchProducer.chainId,
        privateKey: utils.rm0x(pk),
        sender: assetBenchProducer.account.address,
      }),
      OPTIONS: JSON.stringify(opts),
    });

    worker.on('exit', () => console.log('worker done'));
    worker.on('error', (err) => console.log('worker error', err));

    return worker;
  });

  return runMain(assetBenchProducer, workers, opts);
}
